### ISSUE TITLE
Unreadable io::Error debug strings

### ISSUE THREAD

**Issue by liigo** [rust-lang/rust#34318]
While trying to open a non-existing file (on Windows 10, Chinese):

```rust
use std::fs::File;
fn main() {
    File::open("no-such-file.rs").unwrap();
}
```

It panicked with a message:

```
thread '<main>' panicked at 'called `Result::unwrap()` on an `Err` value: Error { repr: Os { code: 2, message: "\u{7cfb}\u{7edf}\u{627e}\u{4e0d}\u{5230}\u{6307}\u{5b9a}\u{7684}\u{6587}\u{4ef6}\u{3002}" } }', ../src/libcore\result.rs:746
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

The OS native error message here is **unreadable** for human being.

The message (`"\u{7cfb}\u{7edf}\u{627e}\u{4e0d}\u{5230}..."`) is the _escaped_ result of "No such file or directory" in Chinese ("系统找不到指定的文件")。[`impl fmt::Debug for str`](http://doc.rust-lang.org/stable/src/core/up/src/libcore/fmt/mod.rs.html#1350-1368) do this escape.

I can read Chinese, which is my mother tongue, but I can't read `\u{7cfb}...`.

Possible solutions:
- Don't use local language (Chinese here) OS error string, use English instead.
- Change `impl fmt::Debug for str` to not escape most Unicode chars. (breaking-change?)

---

```
fn main() {
    let s = "Hello δεΣ❤ 日月";
    println!("{}", s);   // prints "Hello δεΣ❤ 日月"
    println!("{:?}", s); // prints "Hello \u{3b4}\u{3b5}\u{3a3}\u{2764} \u{65e5}\u{6708}"
}
```

[Run it on play.rust-lang.org](https://is.gd/T7lmqG)

Is `impl fmt::Debug for str` really friendly enough for _debug_ purpose? Is there possibility that we change its implementation as similar as `impl fmt::Display for str`?

**Comment by retep998** [comment]
I think printing escapes is very useful for debugging strings, since often consoles will be either incapable of displaying the character (the WIndows console is notorious for this), or it is something like the wrong kind of whitespace slipping in which can be _really_ hard to notice otherwise. I feel any solution for this issue should be targeted at `Debug` for `io::Error` specifically and not strings in general.

**Comment by bluss** [comment]
I'd prefer to output printable characters as they are, escaping is inconvenient for everything that's not in English.

**Comment by liigo** [comment]
> since often consoles will be either incapable of displaying the character (the WIndows console is notorious for this)

Which morden console don't support printing Unicode strings? Do you means that very old one? (Windows 10 console do its job perfectly.)

**Comment by retep998** [comment]
@liigo While the Windows console is capable of storing arbitrary unicode text, its unicode text rendering abilities are far more limited. It only displays unicode glyphs when the font supports it (it doesn't use font fallback) and it also doesn't work for anything outside the BMP, nor does it support combining characters. It's a simple wchar -> glyph mapping. Here is your chinese error string in the Windows 10 console:
![](http://i.imgur.com/AotVWcN.png)
Granted I can easily copy the text from the console and paste it into another program to view it, unless I choose a raster font for the console, in which case the Windows console actually limits itself to u8 instead of u16, completely ruining unicode and the ability to copy paste unicode text, and the only characters you can rely on are ascii.

**Comment by liigo** [comment]
Yes. It's the same when you `println!` something.  I don't think we should implement `println!` to escape almost all unicode characters. If the (old) consoles can't display unicode properly, things should be changed to consoles.

**Comment by liigo** [comment]
> Here is your chinese error string in the Windows 10 console:
> ![](https://camo.githubusercontent.com/eb2443c126a6dedb3e0a327db9fd6cd6d3e4c03c/687474703a2f2f692e696d6775722e636f6d2f416f745657634e2e706e67)

For the specific issue, the os error string is return at runtime by os as local language, I'm sure it has full support to be displayed properly by the os itself.

**Comment by liigo** [comment]
This issue is still exists. Could you reopen it? @brson @alexcrichton 

@tbu- (the author of https://github.com/rust-lang/rust/pull/34485)
cc https://github.com/rust-lang/rust/issues/35068

**Comment by liigo** [comment]
Fixed in nightly by #37855 many thans to @tbu- !
